File: Data Structure/queue.c
//funtion:initialize,enqueue,dequeue

#include<stdio.h>
#include<stdlib.h>

typedef struct _node{
        int item;
        struct _node* next;
}node;

typedef struct _queue{
        node* head;
        node* tail;
}queue;


void enqueue(queue* q,int item){
        node* new=(node*)malloc(sizeof(node));
        new->item=item;
        new->next=NULL;
        if(!q->head){
                q->head=q->tail=new;
                return;
        }
        q->tail->next=new;
        q->tail=new;
}


void dequeue(queue* q){
        if(!q->head)
                return;
        node* tmp=q->head;
        printf("%d dequeue\n",tmp->item);
        q->head=tmp->next;
        free(tmp);
}

int main(){
        queue* q=(queue*)malloc(sizeof(queue));
        q->head=q->tail=NULL;
        for(int i=1;i<=5;i++)
                enqueue(q,i);
        for(int j=1;j<=5;j++)
                dequeue(q);
        return 0;
}

File: Data Structure/GO/queue.go
//data:10.22
//path:~/Code/GO
//method:newQueue(),enqueue(data),dequeue(),printQueue()
package main

import (
	"errors"
	"fmt"
)

//type node struct {
//	data interface{}
//	next *node
//}

type queue struct {
	head *node //just a dummy
	tail *node
}

func iniNode(t interface{}) *node {
	return &node{t, nil}
}

func newQueue() *queue {
	h := iniNode(0)
	return &queue{h, nil}
}

func (q *queue) enqueue(t interface{}) {
	n := iniNode(t)
	if q.tail == nil { //an empty queue
		q.head.next = n
		q.tail = n
		return
	}
	q.tail.next = n
	q.tail = n
}

func (q *queue) dequeue() (interface{}, error) {
	if q.head.next == nil {
		return nil, errors.New("empty queue")
	}
	tmp := q.head.next
	q.head.next = tmp.next
	return tmp.data, nil
}

func (q *queue) printQueue() {
	fmt.Println("Head")
	tmp := q.head.next
	for tmp != nil {
		fmt.Println(tmp.data)
		tmp = tmp.next
	}
	fmt.Println("Tail")
	fmt.Println()
}
